src/main/java/org/javarosa/xpath/analysis/ReferencesMainInstanceAnalyzer.java

```diff
@@ -0,0 +1,47 @@
+package org.javarosa.xpath.analysis;
+
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.instance.TreeReference;
+
+/**
+ * Analyzes an XPath expression to determine whether it contains a reference to the main data
+ * instance.
+ *
+ * @author Aliza Stone
+ */
+public class ReferencesMainInstanceAnalyzer extends XPathBooleanAnalyzer {
+
+    private String mainInstanceRoot;
+
+    public ReferencesMainInstanceAnalyzer(String instanceName, EvaluationContext ec) {
+        this(instanceName);
+        setContext(ec);
+    }
+
+    public ReferencesMainInstanceAnalyzer(String instanceName) {
+        super();
+        this.mainInstanceRoot = instanceName;
+    }
+
+    @Override
+    public void doNormalTreeRefAnalysis(TreeReference treeRef) throws AnalysisInvalidException {
+        if (treeRef.getName(0).equals(mainInstanceRoot)) {
```

Comment:
The idea here is that we can provide different 'main' instances?

Reply:
That wasn't exactly what I was thinking, no. I just didn't want to hardcode "data" as the root ref name since that is technically not always the case, even within a form

Reply:
Ah. So I think we dont' need to check the name of the root node directly for these checks. 

I think you can detect this purely by comparing 
```
            t.getContextType() == TreeReference.CONTEXT_ABSOLUTE && t.getInstanceName() == null;
```

for checking "Does this reference the "main" instance (IE: `/`)".

In theory this actually might mis-detect the main instance because
```
instance('external')/data/path/to/node

and

/data/path/to/node
```

have the same `0` name at the treeref level.